ssh server issue workarounds and explanation (#4519)
circleci/circleci-docs: jekyll/_cci2/SSH-server.adoc

= SSH Rerun Architecture in Server
:page-layout: classic-docs
:page-liquid:
:icons: font
:toc: macro
:toc-title:

> This section describes how SSH reruns work within installations of CircleCI Server. This information is designed to help system administrators have an overview of the processes involved to help with considerations when customizations are made to an installation, and to help with debugging in the event of a problem.

Author: ```suggestion
This section describes how SSH reruns work within installations of CircleCI Server. This information is designed to help system administrators have an overview to help with considerations when customizations are made to an installation. This guide is also here to help with debugging in the event of a problem.
```

== Rerunning a Job with SSH
For an overview of the purpose of SSH reruns in CircleCI, see the <<ssh-access-jobs#,Debugging with SSH guide>>. The image below describes the IP addresses and ports used to restart jobs within VMs and containers when a rerun with SSH is requested.

> The default situation for a Server installation with no customizations is for the IP address to correctly configure the rerun to be taken from AWS metadata. If any networking customizations are made to the installation from the default `/etc/circleci/public-ipv4` should be created on each Nomad Client and this should be populated with the public (if applicable) or private IP of the Nomad Client.

Author: ```suggestion
The default situation for a Server installation with no customizations is for the IP address to correctly configure the rerun to be taken from AWS metadata. If any networking customizations are made to the installation from the default, then `/etc/circleci/public-ipv4` should be created on each Nomad Client and this should be populated with the public (if applicable) or private IP of the Nomad Client.
```

= SSH Rerun Architecture in Server
:page-layout: classic-docs
:page-liquid:
:icons: font
:toc: macro
:toc-title:
:sectanchors:

This section describes how SSH reruns work within installations of CircleCI Server. This information is designed to help system administrators have an overview to help with considerations when customizations are made to an installation. This guide is also here to help with debugging in the event of a problem with SSH reruns in an installation.

== Rerunning a Job with SSH
For an overview of the purpose of SSH reruns in CircleCI, see the <<ssh-access-jobs#,Debugging with SSH guide>>. The image below describes the IP addresses and ports used to restart jobs within VMs and containers when a rerun with SSH is requested.

> The default situation for a Server installation with no customizations is for the IP address to correctly configure the rerun to be taken from AWS metadata. If any networking customizations are made to the installation from the default, then `/etc/circleci/public-ipv4` should be created on each Nomad Client and this should be populated with the public (if applicable) or private IP of the Nomad Client.

Reviewer: ```suggestion
The default situation for a Server installation with no customizations is for the `outer` agent to use AWS metadata to select the instance's public IP address.

If any networking customizations are made from the default, that either block the metadata endpoint or make the public IP unsuitable, then creating a file at `/etc/circleci/public-ipv4` can be used as an override. This should be created on each Nomad Client, usually via a boot script, and be populated with the IP address that users should connect on.
```

Author: @glenjamin  Thanks! yeah this definitely makes the situation clearer. The only thing I'd be concerned about is the "usually via a boot script" part. I'd rather give definite instructions, with maybe an actual snippet of what they should be populating that file with. But I guess server admins are at a level where they are going to know what to do here?

Reviewer: Good call, let me grab a decent example that would demonstrate the approach

Reviewer: For example, to use the instance private IP via AWS metadata 

```sh
#!/bin/sh
PRIVATE_IP="$(curl http://169.254.169.254/latest/meta-data/local-ipv4)"
mkdir -p /etc/circleci
echo $PRIVATE_IP | tee /etc/circleci/public-ipv4
```

Author: great, thank you!
